subprojects/core/src/test/groovy/org/gradle/internal/buildevents/BuildExceptionReporterTest.groovy

```diff
@@ -461,6 +465,137 @@ $GET_HELP
 """
     }
 
+    // region Duplicate Exception Branch Filtering
+    def "multi-cause exceptions have branches with identical root causes summarized properly"() {
+        def ultimateCause = new RuntimeException("ultimate cause")
+        def branch1 = new DefaultMultiCauseException("first failure", ultimateCause)
+        def branch2 = new DefaultMultiCauseException("second failure", ultimateCause)
+        Throwable exception = new ContextAwareException(new DefaultLenientConfiguration.ArtifactResolveException("task dependencies", "org:example:1.0", [branch1, branch2]))
+
+        when:
+        reporter.buildFinished(result(exception))
+        print(output.value)
+
+        then:
+        output.value == """
+{failure}FAILURE: {normal}{failure}Build failed with an exception.{normal}
+
+* What went wrong:
+Could not resolve all task dependencies for org:example:1.0.
+{info}> {normal}first failure
+   {info}> {normal}ultimate cause
+{info}> {normal}There was 1 additional failure with the same cause that was not printed.
+
+* Try:
+$STACKTRACE
+$INFO_OR_DEBUG
+$SCAN
+$GET_HELP
+"""
+    }
+
+    def "multi-cause exceptions have branches with identical root causes and additional intermediate failures summarized properly"() {
+        def ultimateCause = new RuntimeException("ultimate cause")
+        def branch1 = new DefaultMultiCauseException("first failure", ultimateCause)
+        def branch2 = new DefaultMultiCauseException("second failure", ultimateCause)
+        def intermediateFailure = new DefaultMultiCauseException("intermediate failure", ultimateCause)
+        def branch3 = new DefaultMultiCauseException("third failure", intermediateFailure)
+        Throwable exception = new ContextAwareException(new DefaultLenientConfiguration.ArtifactResolveException("task dependencies", "org:example:1.0", [branch1, branch2, branch3]))
+
+        when:
+        reporter.buildFinished(result(exception))
+        print(output.value)
+
+        then:
+        output.value == """
+{failure}FAILURE: {normal}{failure}Build failed with an exception.{normal}
+
+* What went wrong:
+Could not resolve all task dependencies for org:example:1.0.
+{info}> {normal}first failure
+   {info}> {normal}ultimate cause
+{info}> {normal}There were 2 additional failures with the same cause that were not printed.
+
+* Try:
+$STACKTRACE
+$INFO_OR_DEBUG
+$SCAN
+$GET_HELP
+"""
+    }
+
+    def "multi-cause exceptions have branches with identical root causes summarized properly when ultimate cause is self-caused"() {
+        def ultimateCause = new RuntimeException("ultimate cause")
+        Field field = Throwable.class.getDeclaredField("cause")
+        field.setAccessible(true)
+        field.set(ultimateCause, ultimateCause)
```

Comment:
It might be easier to just implement getCause and have it return itself.

I *think* Groovy would have let you do this without any reflection too.

Reply:
Groovy worked locally, so I changed to this.  `cause` is private on `Throwable`, and there isn't a `setCause` to override, so I think this is probably the best we can do.  I don't think it matters too much.